fw/access_log.c

```diff
@@ -0,0 +1,430 @@
+	HTTP_METHODS(XX)
+#undef XX
+};
+
+static const struct {
+#	define MAX_HTTP_VERSION_LEN 9
+	char name[MAX_HTTP_VERSION_LEN];
+	u8 len;
+} http_versions[] = {
+#define XX(ver, name) [ver] = { name, sizeof(name) - 1 }
+	XX(TFW_HTTP_VER_09, "HTTP/0.9"),
+	XX(TFW_HTTP_VER_10, "HTTP/1.0"),
+	XX(TFW_HTTP_VER_11, "HTTP/1.1"),
+	XX(TFW_HTTP_VER_20, "HTTP/2.0"),
+#undef XX
+};
+
+
+enum {
+#define IGNORE(...)
+#define ENUM(id) idx_ ## id,
+	ACCESS_LOG_LINE(IGNORE, IGNORE, ENUM)
+#undef ENUM
+#undef IGNORE
+	TRUNCATABLE_FIELDS_COUNT
+};
+
+struct str_ref {
+	const char *data;
+	unsigned len;
+};
+
+/** Copies string to buffer if it is not plain. Returns pointer to the
+ * first unused character in the "planar" buffer. */
+static inline char*
+make_plain(char *p, char *end, TfwStr *src, struct str_ref *dst)
+{
+	if (TFW_STR_PLAIN(src)) {
+		dst->data = src->data;
+		dst->len = src->len;
+		return p;
+	} /* protect from BUG_ON in tfw_str_to_cstr */
+	else if (p < end) {
+		dst->len = tfw_str_to_cstr(src, p, end - p);
+		dst->data = p;
+		return p + dst->len;
+	} else {
+		dst->data = "";
+		dst->len = 0;
+		return p;
+	}
+}
+
+/** Truncates truncatable fields if needed. */
+static void
+process_truncated(TfwStr *in, struct str_ref *out, char *p, char *end, 
+		unsigned used_chars)
+{
+	unsigned i;
+	unsigned total_len = 0;
+	unsigned truncated_count = 0;
+	unsigned max_untruncated_len, buf_avail;
+	if (unlikely(used_chars >= ACCESS_LOG_BUF_SIZE))
+		goto no_buffer_space;
+	/* Compute total length of all strings that can be truncated */
+	for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++)
+		total_len += in[i].len;
+	/* Check if we're on happy path: all strings fit */
+	if (likely(total_len + used_chars < ACCESS_LOG_BUF_SIZE)) {
+		for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++)
+			p = make_plain(p, end, in + i, out + i);
+		return;
+	}
+	/* Unhappy path: evenly distribute available buffer space across all
+	 * strings that do not fit */
+	buf_avail = (ACCESS_LOG_BUF_SIZE - used_chars);
+	max_untruncated_len = buf_avail / TRUNCATABLE_FIELDS_COUNT;
+	for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++) {
+		/* we loose some chars due to string "less than", but
+		 * tfw_str_to_cstr accounts terminating NUL to total buffer
+		 * length and would truncate strings anyways */
+		if (in[i].len < max_untruncated_len)
+			total_len -= in[i].len;
+		else
+			truncated_count++;
+	}
+	max_untruncated_len = buf_avail / truncated_count;
+	if (max_untruncated_len < sizeof("..."))
+		goto no_buffer_space;
+	/* Now tuncate/plainarize strings */
+	for (i = 0; i < TRUNCATABLE_FIELDS_COUNT; i++) {
+		if (in[i].len < max_untruncated_len) {
+			p = make_plain(p, end, in + i, out + i);
+		} else {
+			/* we need only part of the string + "...", so
+			 * we have to enforce string copy */
+			out[i].data = p;
+			out[i].len = max_untruncated_len;
+			tfw_str_to_cstr(in + i, p, max_untruncated_len);
+			memcpy(p + max_untruncated_len - 3, "...", sizeof("...") - 1);
```

Comment:
This is relatively big string copying and we call logger from softirq (traffic processing), so it'd be beneficial to use `memcpy_fast()`